### DISCUSSION TITLE
Imitating Nuxt's asyncData: is serverPrefetch meant to be replaced by data in the options API?

### DISCUSSION THREAD

**Discussion by johnRivs** [frandiox/vite-ssr#46] [Q&A] [answered]
I have a page component using the options API.
```js
<template>
    <div>
        {{ foo }}
    </div>
</template>

<script>
    export default {
        data: () => ({
            foo: 'baz'
        }),

        serverPrefetch() {
            this.foo = 'bar';
        }
    }
</script>
```

From the server, the HTML comes with `bar` but after a fraction of a second, `bar` gets updated to `baz`. Is `serverPrefetch` something to be used exclusively with a global store?

**Comment by frandiox** [comment]
Hi there!

I think this looks correct and is the way SSR generally works. The frontend gets `bar` from the SSR run, but during re-hydration it overrides it with `baz` from the hardcoded `data` in the component.

To make the re-hydration work well for your case, instead of hardcoding `baz` in the code, you would need to read its value from the SSR's `initialState`, and fallback to `baz` if it's not provided.

You can do this in the `beforeMount` hook I think (`this.foo = initialState.foo || 'baz'`), or maybe even in the `data` function itself. The `initialState` is provided in the main hook so you can pass it to your components in the way you prefer (using stores, provide/inject, or just rely on page props).

Also, make sure you modify `initialState` in the SSR run so it is provided later to the frontend for re-hydration. There's more info about that [here](https://github.com/frandiox/vite-ssr/#ssr-initial-state-and-data-fetching), and a code example [here](https://github.com/frandiox/vite-ssr/blob/eec71e423c45c026c476f8ff0578a4832563ab4e/examples/vue/src/pages/Homepage.vue#L63) (it uses Suspense instead of onServerPrefetch but the concept is the same).

I think I should make guide on common SSR questions like this with examples 🤔 

Hope that helps!

**Reply by johnRivs** [reply]
So let's say I have a page responsible for displaying a user's profile. You say the data should come from `initialState` in the main hook, which I believe is `export default viteSSR(App, { routes }, ({ initialState }) => {`, right? Why does the main hook fetch the data for that page? What if I have 20 pages that need their own data? How come I don't put that call in the relevant component, in this case something like `pages/user.vue`? The main hook is gonna become hella ugly in no time if I just make all the calls from there, even if I separate each call in its own file.

I had the feeling that main hook was something it gets called once and I'd use for more general data, for example, the currently logged in user. Nothing that necessarily has to do with the content of the page itself.

Other stuff I saw in the examples is `router.beforEach` and `serverPrefetch`. `router.beforEach` brings me back to my concern in the first paragraph and `serverPrefetch` makes more sense to me (similar to Quasar's preFetch or Nuxt's asyncData or fetch) but it makes me duplicate every call in every page component, which I believe is a Vue SSR thing rather than a Vite SSR thing.

Is there aboslutely no way to have a) the call to get the content for that page inside the page component and b) not have duplicated calls?

**Reply by frandiox** [reply]
No no, what I meant about the main hook is that you get the `initialState` object there. `vite-ssr` is rather simple and small, it does not enforce you a way to fetch your data, it just gives you that `initialState` object that you can modify in the server and read from it in the browser. You need to provide that object to the place where you want to fetch your data (e.g. in your page components, in your stores, or in your route hooks).

To add some examples, this uses `serverPrefetch()` and `beforeMount()`:

```js
export default defineComponent({
  name: 'Home',
  data() {
    return {
      homeLocalState: null,
    }
  },
  inject: ['initialState'],
  methods: {
    async fetchMyData() {
      if (!import.meta.env.SSR) {
        // Read in browser, in case the initial state has
        // been populated during SSR
        this.homeLocalState = this.initialState.homeLocalState
      }

      if (!this.homeLocalState) {
        // No data: get it fresh from any API.
        // This could be `await (await fetch('https://my-api.com')).json()`
        this.homeLocalState = await new Promise((resolve) =>
          setTimeout(() => resolve('Data from my API'), 500)
        )

        if (import.meta.env.SSR) {
          // Save this data in SSR initial state for hydration later
          this.initialState.homeLocalState = this.homeLocalState
        }
      }
    },
  },
  async serverPrefetch() {
    // This will be called during SSR
    // and will populate the initial state.
    await this.fetchMyData()
  },
  async beforeMount() {
    // This will run in browser
    // and reuse any existing initial state,
    // or get fresh data from API otherwise.
    await this.fetchMyData()
  },
})
```

This assumes you are using `provide('initialState', initialState)` in the vite-ssr main hook to provide the initial state globally.
Instead of that, you could combine `initialState` with a store (see [Pinia example](https://pinia.esm.dev/ssr/#state-hydration)) and then extract this `fetchMyData` logic in a store action. Or any other abstraction you like and that Vue allows .As long as you modify the `initialState` object during SSR and read from it in the browser, it should work.

> router.beforEach brings me back to my concern in the first paragraph 

Well, if you want to fetch data using vue-router, you can also rely on [beforeEnter](https://next.router.vuejs.org/api/#beforeenter) or in-component [beforeRouteEnter](https://next.router.vuejs.org/guide/advanced/navigation-guards.html#in-component-guards), so you have each page data-fetch logic separate.

> but it makes me duplicate every call in every page component, which I believe is a Vue SSR thing rather than a Vite SSR thing

As far as I know, in Vue you are going to need `serverPrefetch` that is called during SSR, and then `beforeMount` or similar that is called during SPA and re-hydration. But the fetch logic can be abstracted in a method or in a store action.

I would also suggest having a look at [Vitedge](https://github.com/frandiox/vitedge) if you want a more opinionated framework that gives you a place to put your data fetching logic. It wraps `vite-ssr` and can be deployed to any Node.js environment like Express.js or a cloud function.

**Reply by johnRivs** [reply]
`serverPrefetch` waits for the promise to be resolved, as expected. However, `beforeMount` doesn't. When I go to the page, the page component will be swapped inmediately, then perform the request and then change the data. How can I make it behave like `serverPrefetch` in that aspect?

**Reply by frandiox** [reply]
The official [Vue SSR guide](https://ssr.vuejs.org/guide/data.html#logic-collocation-with-components) uses `mounted` instead of `beforeMount`, but I'm not sure if that makes any difference.

Perhaps you can make your `beforeMount` check synchronously if the state ready. If it is, then render normally. This should make it work for the first rendering (after SSR).
Otherwise, trigger an async `fetchMyData` and use `v-if` to render a spinner (or any loading state) until the fetching is over. This would happen on further navigation with RouterLink since this would be SPA behavior.

**Reply by johnRivs** [reply]
That's my conclusion after reading the guide and some StackOverflow answers, but it confuses me a little bit cause none of that happens in Quasar or Nuxt. The prefetch fuction will resolve (showing a progress bar in the process), then swap the component and there's no `v-if` in sight. How do you think they do it? 🤔

**Reply by frandiox** [reply]
They probably rely on route guards behind the scenes to fetch data before entering the route (similar to how `beforeEnter` works in Vue Router). On top of that, perhaps some build-time logic to extract the data-fetcher function from Vue components, not sure.

Btw, the data fetching in `beforeEach` can get rather simple by doing something like `async (to, from) => to.meta.state = await (await fetch(to.meta.api)).json()` or something like that, so you just need to define an API endpoint for each page in the routes array. Then, each page will get the result of that `fetch` call as props. Not sure if you want to reconsider that option to avoid dealing with serverPrefetch, beforeMount, etc.
Although this only work for page components. For smaller components you still need Suspense or serverPrefetch.

**Reply by johnRivs** [reply]
How can I access the instance of the page component inside a navigation guard like `beforeEach` or `beforeResolve`? When I log `to.matched[0].instances` I see this in the console but I can't access that data. `to.matched[0].instances.default.asdf` doesn't work.

![](https://i.imgur.com/DBtxrWV.png)

**Reply by frandiox** [reply]
Not sure how to access the instance but if you write in `to.meta.state = { whatever }` it will be passed as props to the page component.

Also, not sure if you saw this reply (for Suspense): https://github.com/frandiox/vite-ssr/discussions/46#discussioncomment-963069

**Reply by johnRivs** [reply]
Yeah, it's a concise approach but it doesn't quite fit my style. Btw, I recognize this has nothing to do with your package anymore so feel free to ignore my experiments 😂 

My (admittedly very picky) goal:
- Interact with data directly, in other words, no `store.something` or `$attrs.something`.
- Use the options API. I believe using `setup` makes the previous point easier but oh well.
- No inject/provide.
- Results from `prefetch` available from the page component.
- No Vuex.

Essentially, [Nuxt's asyncData](https://nuxtjs.org/docs/2.x/features/data-fetching#async-data).

Basically, this is what I have right now.
During SSR:
1 - Fetch whatever's inside the page component's `prefetch`, get whatever's inside the component's `data`.
2 - I set up a flag for later and merge the 2 things from before. Make it the new component's `data`.
3 - Use `initialState` as a vehicle for later.

Hydration:
4 - Since the flag is true, set the component's data to the merged data from earlier (that I stored in `initialState`).

Client navigation:
5 - Skip the previous step since the flag is false. Fetch data for that page.
6 - Merge current component data and `prefetch` result. Use `initialState` as vehicle.
7 - Whatever's in the `prefetch` result, put it in `data`. The reason I don't `component.data = () => ({ ...data, ...result });` again is because it doesn't work perfectly. If I go to `/users` then navigate to `/users?page=4`, replacing the data function again seems to break reactivity. That's why I'm using the `updated` hook.

Now you should understand my question about the instance. I'd like a better way to "inject" the data into the component without having to do weird merges, replace `data` or `updated` hook.

```js
router.beforeResolve(async (to, from, next) => {
    let component = to.matched[0].components.default;

    // [7]
    component.updated = function() {
        for (const key in this.$attrs.page) {
            this[key] = this.$attrs.page[key];
        }
    };

    // [4]
    if (initialState.flag) {
        initialState.flag = false;
        component.data = () => ({ ...initialState.page });

        return next();
    }

    // [1]
    // [5]
    let result = await component.prefetch({ app, router, initialState, to, api }),
        data   = component.data();

    // [2]
    if (import.meta.env.SSR) {
        initialState.flag = true;
        component.data = () => ({ ...data, ...result });
    }

    // [3]
    // [6]
    initialState.page = { ...data, ...result };

    next();
});
```

**Reply by frandiox** [reply]
@johnRivs Interesting. I think I got something that works. There is no instance created during `beforeResolve` so I think we are forced to override a component hook. This relies on `created`, which runs both in server and client:

```js
  router.beforeResolve(async (to) => {
    // Maybe this should be last matched to support nested routes?
    const component = to.matched[0].components.default

    if (component.prefetch) {
      // Wrap Created hook
      const originalCreatedHook = component.created
      if (!originalCreatedHook || !originalCreatedHook._prefetchWrap) {
        component.created = function (...args) {
          // Merge data in component instance
          Object.assign(this, initialState[to.name] || {})

          if (originalCreatedHook) {
            originalCreatedHook(...args)
          }
        }

        // Mark it to avoid further wrapping
        component.created._prefetchWrap = true
      }

      if (!initialState[to.name]) {
        // Fetch data and save for this route
        initialState[to.name] = await component.prefetch(context)
      }
    }
  }
```

This is basically using `initialState` as a central store for fetched data, even in browser.
Would something like this fit your needs?

**Reply by johnRivs** [reply]
I like it. Clever and compact. 

There's one issue I can think of. If I navigate from `/users` to `/users?page=7` created won't be enough. Updated covers that but it doesn't run during SSR.

**Reply by johnRivs** [reply]
I think I got it 😭 Also works when navigating to the same page (e.g. pagination) and can be adjusted if you want to support nested routes. Thanks for going back and forth with me.
```js
router.beforeResolve(async (to, from, next) => {
    let component = to.matched[0].components.default,
        instance  = to.matched[0].instances.default;

    if (! component.prefetch || ! component) return next();

    to.meta = {};

    if (! from.href && isClient) {
        component.data = () => ({ ...initialState.pageData });

        return next();
    }

    let result = await component.prefetch({ app, router, initialState, to, api }),
        data   = ! instance && component.data?.();

    instance ? Object.assign(instance, result) : component.data = () => ({ ...data, ...result });

    if (! isClient) initialState.pageData = { ...data, ...result };

    next();
});
```

Usage:
```vue
<template>
    <div>
        home page {{ asdf }} {{ qwer }}
    </div>
</template>

<script>
    export default {
        data: () => ({
            asdf: 'zxcv'
        }),

        prefetch({ api, to }) {
            return new Promise((resolve) => {
                setTimeout(() => resolve({ qwer: +new Date }), 100)
            });
        }
    }
</script>
```

**Reply by frandiox** [reply]
@johnRivs That's great! I wonder if we should add an optional module to vite-ssr to implement this 🤔

**Reply by lockinmarv** [reply]
Not sure if I missed something, but is it possible to do the same thing for components? Currently I need to fetch data in the page component, wait with `<Suspense>` and pass it down to the components. Nuxt's `asyncData` is for pages where `fetch` can be called in components too. What would be the solution for it?

I thought the components should do the data fetching and not the page. If I have 3 different components that fetch from 3 different data sources, would I do that in the page component and pass the result to each component? 

Okay nvm, I thought Nuxt handles it differently and fetches the data in the component before the route is visible. But the behavior is the same.

**Comment by johnRivs** [comment]
Ahh, I see what you mean now. I thought you were saying to fetch the data first, then put it into initialState and then pass it to the component (although that's possible too I guess, for the logged in user example). My bad.

> Well, if you want to fetch data using vue-router, you can also rely on beforeEnter or in-component beforeRouteEnter, so you have each page data-fetch logic separate.

Yeah, I didn't mention those cause of the lack of `this` which may or may not end up being a problem.

Vitedge looks really cool, I'm just not familiar with that type of workflow and deployment, but with your response I should be able to keep going with vite-ssr. Thanks for your time!

**Comment by lockinmarv** [comment]
Would it be possible to provide some hooks like Nuxt's `fetch` or `asyncData` so you can fetch and return data in the component (pages)? `router.beforeEach` is kinda okay but imo it would be much better to fetch the data directly in the components. I use Composition API which doesn't work nicely with `beforeRouterEnter`.

**Reply by frandiox** [reply]
Possible yes, but I would prefer to keep `vite-ssr` as a simple library to do SSR and leave sugar-syntax fetching to each framework (React/Vue/etc) or any other library wrapping `vite-ssr` (like [Vitedge](https://github.com/frandiox/vitedge)).

In composition API, the simplest way is probably using `async setup` + `<Suspense>`. That will stop the rendering in both browser and server until the data is fetched. There's an [example here](https://github.com/frandiox/vite-ssr/blob/master/examples/vue/src/pages/Homepage.vue#L28).

You can easily extract that logic into a hook and just do `async setup() {  const state = await useFetchData() }` from your components:

```js
import { inject, ref } from 'vue'
import { useRoute } from 'vue-router'

export async function useFetchData() {
  const { name } = useRoute() // this is just a unique key
  const initialState = inject('initialState')
  const state = ref(initialState[name] || null)

  if (!state.value) {
    state.value = await (await fetch('https://my-api.com?page=' + name)).json()

    if (import.meta.env.SSR) {
      initialState[name] = state.value
    }
  }

  return state
}
```

You can modify that to accept parameters instead of relying on Route's name or anything you need.

**Reply by lockinmarv** [reply]
Oh wow, simple as that? I will try that, thanks for your feedback again. It's slowly become clearer. 😄

**Comment by johnRivs** [comment, marked as answer]
End result: https://github.com/frandiox/vite-ssr/discussions/46#discussioncomment-988827
